### PR TITLE
Add a buffer to the export ostream

### PR DESCRIPTION
**Pull request by paulhuggett**
The main reason for poor performance of the export code is that each write to the export::ostream results in a call to one of the standard library fputc(), fputs(), etc. functions. On macOS at least, these take a lock on the output stream: continually locking and unlocking the underlying mutex is slow.

This change replaces that very simple implementation with one that buffers its output before finally calling fwrite(). We also add an ostream subclass which writes to a std::string (a la std::ostringstream). This lets me eliminate a lot of the template noise which was necessary to enable the unit tests to use the standard library ostream which the real program used export::ostream.

### CODE REVIEW

**Comment by MaggieYingYi** [comment]
Could the following function in the `export_emit.hpp` file remove the `OSStream` and use `ostream_base`?

```
template <typename OSStream, typename ElementType, std::ptrdiff_t Extent>
        void write_span (OSStream & os, gsl::span<ElementType, Extent> const & sp)
```

File: lib/exchange/export_compilation.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+#include "pstore/exchange/export_ostream.hpp"
+
+using pstore::exchange::export_ns::ostream_base;
+
+namespace pstore {
+    namespace repo {
+
+#define X(a)                                                                                       \
+    case repo::linkage::a: return os << #a;
+        ostream_base & operator<< (ostream_base & os, linkage const l);
```
Comment (MaggieYingYi) [inline lib/exchange/export_compilation.cpp:34]: Next line is the definition of `operator<<`. Why need this line?

File: lib/exchange/export_compilation.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        ostream_base & operator<< (ostream_base & os, linkage const l) {
+            switch (l) { PSTORE_REPO_LINKAGES }
+            return os << "unknown";
+        }
+#undef X
+
+        ostream_base & operator<< (ostream_base & os, visibility const v);
```
Comment (MaggieYingYi) [inline lib/exchange/export_compilation.cpp:41]: Next line is the definition of `operator<<`. Why need this line?

**Review by MaggieYingYi — COMMENTED** [review]
Some functions in `export_emit.hpp` and `export_section.hpp` aren’t eliminated the template. What is the reason? Do you measure the performance improvement before and after this change?

**Comment by paulhuggett** [comment]
> Some functions in `export_emit.hpp` and `export_section.hpp` aren’t eliminated the template. What is the reason? Do you measure the performance improvement before and after this change?

Do you mean that I haven’t removed the OStream template argument in these case? If so, then I didn’t have an explicit goal of eliminating them in all cases. The template argument makes no difference to performance (that comes from the ostream implementation itself). The motivation for that is just to simplify things.

In some cases — specifically, the higher order functions — the template can’t be eliminated because the caller may wish to use the concrete subclass in all of its interfaces.

File: lib/exchange/export_compilation.cpp (hunk 1 above)
Comment (paulhuggett) [reply lib/exchange/export_compilation.cpp:34]: It avoids a compiler warning that an external function doesn’t have a declaration.

File: lib/exchange/export_compilation.cpp (hunk 2 above)
Comment (paulhuggett) [reply lib/exchange/export_compilation.cpp:41]: It avoids a compiler warning that an external function doesn’t have a declaration.

Comment (paulhuggett) [reply lib/exchange/export_compilation.cpp:41]: As above.

**Review by MaggieYingYi — APPROVED** [review]
LGTM!

**Pull request merged**